ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -120,6 +120,14 @@ class PutOperation {
   private boolean isEncryptionEnabled;
   private final QuotaChargeCallback quotaChargeCallback;
 
+  // Whether the entire blob (all chunks) are compressible.  Default is null for undecided.
+  // The value is set based on blobProperties and compression service configuration.
+  // For example, if BlobProperties.ContentEncoding specified, then the entire blob will
+  // not be compressed by setting this field to FALSE.
+  // This field is set once when inserting the first chunk based on blob properties.
+  // All chunk compression will check this field and skip compression if this value FALSE.
+  private Boolean isBlobCompressible;
```

Comment:
Can we also check the "x-ambry-content-encoding". Customer will set it to gzip if the blob has been pre-compressed. So we don't need to compress twice. But during decompress stage, we should not decompress it. Customer will handle the decompression.

Reply:
Sure thing.  Let me make that change.

Reply:
You can also have a follow up pr if you think this pr is huge.

Reply:
According to RestUtil.java, the BlobProperties.contentEncoding is obtained from the "x-ambry-content-encoding" header.  So I believe we are currently doing that already.

    public final static String AMBRY_CONTENT_ENCODING = "x-ambry-content-encoding";

public static BlobProperties buildBlobProperties(Map<String, Object> args) throws RestServiceException {
    Account account = getAccountFromArgs(args);
    Container container = getContainerFromArgs(args);
    String serviceId = getHeader(args, Headers.SERVICE_ID, true);
    String contentType = getHeader(args, Headers.AMBRY_CONTENT_TYPE, true);
    String ownerId = getHeader(args, Headers.OWNER_ID, false);
    String externalAssetTag = getHeader(args, Headers.EXTERNAL_ASSET_TAG, false);
    **String contentEncoding = getHeader(args, Headers.AMBRY_CONTENT_ENCODING, false);**
    String filename = getHeader(args, Headers.AMBRY_FILENAME, false);

    long ttl = getTtlFromRequestHeader(args);

    // This field should not matter on newly created blobs, because all privacy/cacheability decisions should be made
    // based on the container properties and ACLs. For now, BlobProperties still includes this field, though.
    boolean isPrivate = !container.isCacheable();
    return new BlobProperties(-1, serviceId, ownerId, contentType, isPrivate, ttl, account.getId(), container.getId(),
        container.isEncrypted(), externalAssetTag, contentEncoding, filename);
  }

Reply:
yeah, you are right, I missed this part. I think it's good!
`if (isSkipWithContentEncoding && !Utils.isNullOrEmpty(contentEncoding))`